store/tikv/latch/latch.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package latch
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spaolacci/murmur3"
+)
+
+// Latch stores a key's waiting transactions
+type Latch struct {
+	waiting      []uint64
+	lastCommitTs uint64
+	lock         sync.Mutex
+}
+
+// acquire tries to get current key's lock for the transaction with startTs
+// acquire is true when success
+// timeout is true when the startTs is already timeout
+// newWait is true when current transaction is new for the current latch
+func (l *Latch) acquire(startTs uint64) (acquire, timeout, newWait bool) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	timeout = startTs <= l.lastCommitTs
+	if timeout {
+		return
+	}
+	if len(l.waiting) == 0 || l.waiting[0] != startTs {
+		l.waiting = append(l.waiting, startTs)
+		newWait = true
+	}
+
+	acquire = l.waiting[0] == startTs
+	return
+}
+
+// release releases the transaction with startTs and commitTs from current latch.
+// isEmpty is true when the waiting queue is empty after release current transaction,
+// otherwise return the front transaction in queue.
+func (l *Latch) release(startTs uint64, commitTs uint64) (isEmpty bool, front uint64) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	if startTs != l.waiting[0] {
+		panic(fmt.Sprintf("invalid front ts %v, latch:%+v", startTs, l))
+	}
+	// for rollback or timeout, the commitTs maybe zero
+	if commitTs > l.lastCommitTs {
```

Comment:
if you want to avoid assign zero to `lastCommitTs`, why not just explicitly check `commitTs` is zero or not? If the value is not zero, then assign `commitTs` to `lastCommitTs`.